### ISSUE TITLE
Segmentation fault at consensus step

### ISSUE THREAD

**Issue by sscalabrin** [marbl/canu#1845]
Hello,

here the command:
canu -p xxx -d xxx_raw_minRead10k_ovl3k gnuplot=/u/iga/sscalabrin/.conda/envs/canu/bin/gnuplot java=/u/iga/sscalabrin/.conda/envs/canu/bin/java useGrid=remote gridEngine="pbspro" gridEngineResourceOption="-q thin -l select=1:ncpus=12:mem=300g -l walltime=96:00:00" genomeSize=1.5g -nanopore PATH/*gz rawErrorRate=0.5 correctedErrorRate=0.144 minReadLength=10000 minOverlapLength=3000 corOutCoverage=200 ovlMerDistinct=0.975 corMhapSensitivity=high saveReads=True "batOptions=-dg 3 -db 3 -dr 1 -ca 500 -cp 50"

here the error:
```
Found perl:
   /bin/perl
   This is perl 5, version 16, subversion 3 (v5.16.3) built for x86_64-linux-thread-multi

Found java:
   /u/iga/sscalabrin/.conda/envs/newcanu/bin/java
   openjdk version "1.8.0_152-release"

Found canu:
   /large/___HOME___/iga/sscalabrin/.conda/envs/newcanu/bin/canu
   canu branch HEAD +0 changes (r10117 5638f7d9a5379373310ab62c28aa0cdbd864722d)

Running job 11 based on command line options.
-- Using seqFile '../xxx.ctgStore/partition.0011'.
-- Opening tigStore '../xxx.ctgStore' version 1.
-- Opening output results file './ctgcns/0011.cns.WORKING'.
--
-- Computing consensus for b=0 to e=4901675 with errorRate 0.1440 (max 0.4000) and minimum overlap 500
--
Loading corrected reads from seqFile '../xxx.ctgStore/partition.0011'

Failed with 'Segmentation fault'; backtrace (libbacktrace):
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
PATH/xxx_raw_minRead10k_ovl3k/unitigging/5-consensus_temp/consensus.sh: line 106:  7345 Segmentation fault      (core dumped) $bin/utgcns -R ../xxx.${tag}Store/partition.$jobid -T ../xxx.${tag}Store 1 -P $jobid -O ./${tag}cns/$jobid.cns.WORKING -maxcoverage 40 -e 0.144 -pbdagcon -edlib -threads 1
```

I had the issue with Canu v2.0. I installed the v2.1, same issue
I tried https://github.com/marbl/canu/files/5162616/utgcns.gz from https://github.com/marbl/canu/issues/1780 but it did not work either

xxx.ctgStore, xxx.utgStore, xxx.ovlStore are 57, 39, 34 Gb, respectively.

Only partition 11 failed, out of 52.

Any idea?
Best regards,
Simone Scalabrin

**Comment by skoren** [comment]
The consensus issue wasn't fixed until 2.1.1 but it looks like you're actually running 2.1.1 (at least the tag says this). It looks like your compiled version doesn't have the stack trace enabled so it's not clear if it's the same issue as #1780 or something else. Are you able to share the data (we'd need the ctgStore and seqStores)?

**Comment by sscalabrin** [comment]
Dear Sergey,

indeed you are right. I used v2.1.1.
I downloaded it from bioconda (https://anaconda.org/bioconda/canu).

I uploaded you the two directories here:
https://support.igatech.it/delivery/
Username : koren_sergey
Password : canu2020

Thank you so much for looking into it!

The only suspect I have is that I run different steps in different servers but with the same version of Canu, 2.0
In particular, I run some of the alignments in a cluster and some in multiple standalone machines. In the latter case, commands issued were something like:
./overlap.sh 337 >utgovl_xxx.oLOCAL.000337.out 2>&1

I never had problems with this way to compute the assembly with different datasets.

Best regards,
Simone

**Comment by skoren** [comment]
Thanks for uploading the data to our FTP, I was able to download it from there. It looks like we're getting an out of bounds error which is relatively easy to fix in the code, change `src/utgcns/unitigConsensus.C from:
```
 704     // let the band increase without increasing error rate for a while, if we give up increase error rate
 705     // and reset bnad
 706     if (ii != 0 && ii % NUM_BANDS == 0) {
 707        bandErrRate += errorRate / ERROR_RATE_FACTOR;
 708        tigbgn = max((int32)0,      (int32)floor(utgpos.min() - padding));
 709        tigend = min((int32)tiglen, (int32)floor(utgpos.max() + padding));
 710     }
```
to:
```
 704     // let the band increase without increasing error rate for a while, if we give up increase error rate
 705     // and reset bnad
 706     if (ii != 0 && ii % NUM_BANDS == 0) {
 707        bandErrRate += errorRate / ERROR_RATE_FACTOR;
 708        tigbgn = max((int32)0,      (int32)floor(utgpos.min() - padding));
 709        tigend = min((int32)tiglen, (int32)floor(utgpos.max() + padding));
 710     }
 711 
 712   if (tigend < tigbgn) {
 713     fprintf(stderr, "alignEdLib()-- WARNING: tigbgn %d > tigend %d - tiglen %d utgpos %d-%d padding %d\n",
 714             tigbgn, tigend, tiglen, utgpos.min(), utgpos.max(), padding);
 715     // try to align it to full
 716     tigbgn = 0;
 717     tigend = utgpos.max();
 718     fprintf(stderr, "alignEdLib()-- WARNING: updated tigbgn %d > tigend %d - tiglen %d utgpos %d-%d padding %d\n",
 719             tigbgn, tigend, tiglen, utgpos.min(), utgpos.max(), padding);
 720   }
 721   assert(tigend > tigbgn);
```

The interesting thing is that this partition seems to have several tigs which are much larger than they should be, I am guessing this happened in bogart before consensus. How big is your ovlStore folder? It may be interesting on our end to reproduce that part as well but we'd need the ovlStore.

**Comment by sscalabrin** [comment]
Thank you so much!
Tomorrow I will try the patch.
In the meanwhile, I am uploading you the ovlStore as a single tarball.
34Gb, it should last one more hour.

I gotta go, its 10pm here.
Best,
Simone

ps I don't think this will help you but I am assembling the olive tree genome, a very heterozygous plant, about 3%, e.g. one base every 30bp. Haploid genome size 1.5Gb. I'd like to have a haplotype resolved sequence. A first assembly gave me 2.6Gb.

Simone Scalabrin

 [cid:dot_only_140h_ee82ae65-db65-493d-893c-e4a1ae0d8aa9.png]


Agrigenomics & Research Projects Lead
IGA Technology Services
a: Via J.Linussio, 51, 33100 Udine, Italy
w: www.igatechnology.com<http://www.igatechnology.com/>
[cid:FB-f-Logo__blue_29_fefc7735-e8df-4d1b-944b-e82473bb9aac.png]<https://facebook.com/IGATechnology>   [cid:Twitter_logo_blue_32_cdd3ea58-fd8e-4bbc-8987-015575df7b2d.png] <https://twitter.com/igatechnology>    [cid:linkedin_sig_8e470841-b4f0-4907-b7cb-e723a1440c9f.png] <https://linkedin.com/company/iga-technology-services>

________________________________

**Comment by sscalabrin** [comment]
Hi Sergey,

I compiled the unreleased version after applying the modification in the code. It does compile, everything looks ok but I obtain a similar error (if not the same):

cd /PATH/assembly/canu; canu -p Leccino -d Leccino_raw_minRead10k_ovl3k gnuplot=/u/iga/sscalabrin/.conda/envs/compile_canu/bin/gnuplot java=/u/iga/sscalabrin/.conda/envs/compile_canu/bin/java useGrid=remote gridEngine="pbspro" gridEngineResourceOption="-q fat -l select=1:ncpus=8:mem=30g -l walltime=96:00:00" genomeSize=1.5g -nanopore /u/iga/sscalabrin/scratch/ID1384_OlgenomePATH/lanes/sequences/promethion/*gz rawErrorRate=0.5 correctedErrorRate=0.144 minReadLength=10000 minOverlapLength=3000 corOutCoverage=200 ovlMerDistinct=0.975 corMhapSensitivity=high saveReads=True "batOptions=-dg 3 -db 3 -dr 1 -ca 500 -cp 50"

./consensus.jobSubmit-01.sh gets submitted in the cluster and runs 10 seconds:

Found perl:
   /bin/perl
   This is perl 5, version 16, subversion 3 (v5.16.3) built for x86_64-linux-thread-multi

Found java:
   /u/iga/sscalabrin/.conda/envs/compile_canu/bin/java
   openjdk version "1.8.0_92"

Found canu:
   /large/___HOME___/iga/sscalabrin/bin/canu/build/bin/canu
   canu snapshot v2.2-development +80 changes (r10189 aa43bb570086c1e653ac0df66afd67fb9b4d0a19)

Running job 11 based on command line options.
-- Using seqFile '../Leccino.ctgStore/partition.0011'.
-- Opening tigStore '../Leccino.ctgStore' version 1.
-- Opening output results file './ctgcns/0011.cns.WORKING'.
--
-- Computing consensus for b=0 to e=4901675 with errorRate 0.1440 (max 0.4000) and minimum overlap 500
--
Loading corrected reads from seqFile '../Leccino.ctgStore/partition.0011'

Failed with 'Segmentation fault'; backtrace (libbacktrace):
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
(null)::0 in (null)()
/PATH/Leccino_raw_minRead10k_ovl3k/unitigging/5-consensus/consensus.sh: line 96: 19797 Segmentation fault      (core dumped) $bin/utgcns -R ../Leccino.ctgStore/partition.$jobid -T ../Leccino.ctgStore 1 -P $jobid -O ./ctgcns/$jobid.cns.WORKING -maxcoverage 40 -e 0.144 -pbdagcon -edlib -threads 8

**Comment by skoren** [comment]
Not sure why the patch isn't working for you. I attached the output of the job (the ins file) and the compiled utgcns binary which hopefully would work on your system as well.

[utgcns.gz](https://github.com/marbl/canu/files/5623655/utgcns.gz)
[0011.cns.gz](https://github.com/marbl/canu/files/5623654/0011.cns.gz)

**Comment by sscalabrin** [comment]
utgcns still not working, probably a linking issue:
utgcns: error while loading shared libraries: libgsl.so.23: cannot open shared object file: No such file or directory
I'll check that later

while cns file is working, the assembly is running for the last two hours

**Comment by skoren** [comment]
Yes, the error from utgcns looks like a linking issue because we have a newer OS than you're running it on. I committed the fix to the repo so you could re-build to get the fix as well. Since it went past the stage and I was able to run locally, I'm going to close the issue. If you encounter errors at later stages open a new issue.

**Comment by ASLeonard** [comment]
I had the same issue with certain jobs failing *ctgcns*, and the fix you committed in af771ef solved this issue. I know the issue is already closed, but just some extra confirmation.
